node/src/components/chain_synchronizer.rs

```diff
@@ -308,24 +346,33 @@ impl ChainSynchronizer {
                 );
 
                 let initial_pre_state = ExecutionPreState::new(
-                    upgrade_block_header.height() + 1,
+                    switch_block_header_before_upgrade.height() + 1,
                     post_state_hash,
-                    upgrade_block_header.hash(self.config.verifiable_chunked_hash_activation()),
-                    upgrade_block_header.accumulated_seed(),
+                    switch_block_header_before_upgrade
+                        .hash(self.config.verifiable_chunked_hash_activation()),
+                    switch_block_header_before_upgrade.accumulated_seed(),
                 );
                 let finalized_block = FinalizedBlock::new(
                     BlockPayload::default(),
                     Some(EraReport::default()),
-                    upgrade_block_header.timestamp(),
-                    upgrade_block_header.next_block_era_id(),
+                    switch_block_header_before_upgrade.timestamp(),
+                    switch_block_header_before_upgrade.next_block_era_id(),
                     initial_pre_state.next_block_height(),
                     PublicKey::System,
                 );
+                // If this is an emergency upgrade, we don't need to pass the switch block from just
+                // before the upgrade, as it's only used to derive the list of validators to sign
+                // the immediate switch block, and for an emergency upgrade that list is the same as
+                // its own `next_era_validators` collection.
+                let maybe_switch_block_header_before_upgrade =
+                    (!is_emergency_upgrade).then(|| switch_block_header_before_upgrade);
```

Comment:
Should we determine that inside of `execute_immediate_switch_block` instead of passing it in?

Reply:
We call `execute_immediate_switch_block` from `handle_commit_genesis_result` too though, and in that path we don't have a previous switch block available to pass in.